src/main/java/frc/robot/subsystems/Vision/VisionHelper.java

```diff
@@ -0,0 +1,423 @@
+    double[] minAreaRectCornersY = table.get("min area rect corners y " + name, new double[4]);
+
+    List<TargetCorner> detectedCorners = new ArrayList<>();
+    List<TargetCorner> minAreaRectCorners = new ArrayList<>();
+
+    for (int i = 0; i < 4; i++) {
+      detectedCorners.add(new TargetCorner(detectedCornersX[i], detectedCornersY[i]));
+      minAreaRectCorners.add(new TargetCorner(minAreaRectCornersX[i], minAreaRectCornersY[i]));
+    }
+    Transform3d pose = getLoggedTransform3d(translation, rotation);
+    Transform3d altPose = getLoggedTransform3d(altTranslation, altRotation);
+    return (new PhotonTrackedTarget(
+        table.get("yaw " + name, -1),
+        table.get("pitch " + name, -1),
+        table.get("area " + name, -1),
+        table.get("skew " + name, -1),
+        (int) (table.get("fiducial id " + name, -1)),
+        pose,
+        altPose,
+        table.get("pose ambiguity " + name, -1),
+        minAreaRectCorners,
+        detectedCorners));
+  }
+  /**
+   * Poll data from the configured cameras and update the estimated position of the robot. Returns
+   * empty if there are no cameras set or no targets were found from the cameras.
+   *
+   * @return an EstimatedRobotPose with an estimated pose, the timestamp, and targets used to create
+   *     the estimate
+   */
+  public static Optional<EstimatedRobotPose> update(
+      PhotonPipelineResult result,
+      AprilTagFieldLayout fieldTags,
+      PoseStrategy primaryStrategy,
+      PoseStrategy multiTagFallbackStrategy) {
+
+    double poseCacheTimestampSeconds = -1;
+
+    // Time in the past -- give up, since the following if expects times > 0
+    if (result.getTimestampSeconds() < 0) {
+      return Optional.empty();
+    }
+
+    // If the pose cache timestamp was set, and the result is from the same timestamp, return an
+    // empty result
+    // if (poseCacheTimestampSeconds > 0
+    //     && Math.abs(poseCacheTimestampSeconds - result.getTimestampSeconds()) < 1e-6) {
+    //   return Optional.empty();
+    // }
+
+    // Remember the timestamp of the current result used
+    poseCacheTimestampSeconds = result.getTimestampSeconds();
+
+    // If no targets seen, trivial case -- return empty result
+    if (!result.hasTargets()) {
+      return Optional.empty();
+    }
+
+    Optional<EstimatedRobotPose> estimatedPose;
+    switch (primaryStrategy) {
+      case LOWEST_AMBIGUITY:
+        estimatedPose = lowestAmbiguityStrategy(result, fieldTags);
+        break;
+      case MULTI_TAG_PNP_ON_COPROCESSOR:
+        estimatedPose =
+            multiTagPNPStrategy(result, fieldTags, primaryStrategy, multiTagFallbackStrategy);
+        break;
+      default:
+        DriverStation.reportError(
+            "[PhotonPoseEstimator] Unknown Position Estimation Strategy!", false);
+        return Optional.empty();
+    }
+
+    return estimatedPose;
+  }
+
+  private static Optional<EstimatedRobotPose> multiTagPNPStrategy(
```

Comment:
clarify that this is pnp on rio, maybe make sure it uses pnp on coproc when available